ddtrace/contrib/internal/ray/span_manager.py

```diff
@@ -0,0 +1,241 @@
+import atexit
+from contextlib import contextmanager
+from itertools import chain
+import threading
+import time
+
+from ddtrace import config
+from ddtrace import tracer
+from ddtrace._trace.span import Span
+from ddtrace.constants import SPAN_KIND
+from ddtrace.ext import SpanKind
+
+from .utils import _inject_ray_span_tags_and_metrics
+
+
+@contextmanager
+def long_running_ray_span(span_name, service, span_type, child_of=None, activate=True):
+    """Context manager that handles Ray span creation and long-running span lifecycle"""
+    with tracer.start_span(
+        name=span_name, service=service, span_type=span_type, child_of=child_of, activate=activate
+    ) as span:
+        span.set_tag_str(SPAN_KIND, SpanKind.CONSUMER)
+        _inject_ray_span_tags_and_metrics(span)
+        start_long_running_span(span)
+
+        try:
+            yield span
+        except BaseException as e:
+            span.set_exc_info(type(e), e, e.__traceback__)
+            raise
+        finally:
+            stop_long_running_span(span)
+
+
+class RaySpanManager:
+    def __init__(self):
+        self._timers = {}
+        # {submission_id: {(trace_id, span_id): Span}}
+        self._job_spans = {}
+        # {submission_id: (Span)}
+        self._root_spans = {}
+        self._lock = threading.Lock()
+        self._is_shutting_down = False
+
+        # Register cleanup on process exit
+        atexit.register(self.cleanup_on_exit)
+
+    def _get_submission_id(self, span):
+        submission_id = span.get_tag("ray.submission_id")
+        return submission_id if submission_id else None
+
+    def cleanup_on_exit(self):
+        """Clean up all resources when the process exits."""
+        with self._lock:
+            self._is_shutting_down = True
+
+            for _, timer in list(self._timers.items()):
+                timer.cancel()
+
+            # Check if we have any spans to flush before exiting
+            spans_to_close = []
+            for spans_dict in self._job_spans.values():
+                spans_to_close.extend(list(spans_dict.values()))
+
+            self._timers.clear()
+            self._job_spans.clear()
+            self._root_spans.clear()
+
+        for span in spans_to_close:
+            self._finish_span(span)
+
+    def _emit_partial_span(self, span):
+        partial_version = time.time_ns()
```

Comment:
A cleaner approach here could be to add a tag to spans that are allowed to be long running. Then turn the SpanAggregstor into a periodic thread that flushes spans all finished spans every X milliseconds and flushes long running spans every Y seconds. We can do this in a follow up. 

This way we limit impact of this mechanism to select integration spans and can slowly expand it to other integrations and maybe manual spans.